Support sbt 1.4 virtual files for compilation and config errors

mkurz opened the pull request:
Follow up to #10648, but this time it's about compilation/config errors.
(Please see the description of #10648 how this sbt 1.4 virtual file stuff works.)

Along the way I also fixed #10710, which I ran into when working on the scripted test. Like described in #10710, an invalid `conf/messages` file never displayed a nice error page, which however was intended because a `PlayException` get thrown (including the line number of the messages file etc.)

The scripted test basically tests (hopefully) all possible compilation and config errors that can occur and that Play handles.

Backport: #10712

mkurz commented on core/play/src/main/scala/play/api/i18n/Messages.scala:
```diff
@@ -89,7 +89,7 @@ object Messages extends MessagesImplicits {
       messageSourceName: String
   ): Either[PlayException.ExceptionSource, Map[String, String]] = {
-    new Messages.MessagesParser(messageSource, "").parse.right.map { messages =>
+    new Messages.MessagesParser(messageSource, messageSourceName).parse.right.map { messages =>
```
I have no idea why in 2015 Rich passed an empty string, I just couldn't find any sense in it:
https://github.com/playframework/playframework/pull/4588/files#diff-5d232f7f329d188def0922b9f06ecf2d3d19844f9c6012bb007bd2d33e936fb7R240
I think that was mistake because that change in the 2015 pull request is not related at all to messages. I guess Rich did some testing and accidentally commited the empty string...

marcospereira replied:
Yeah, I think you are right. Good to fix it here too. :-)

mkurz commented on core/play/src/main/scala/play/api/i18n/Messages.scala:
```diff
@@ -595,7 +595,11 @@ class DefaultMessagesApiProvider @Inject() (
+            Messages.UrlMessageSource(messageFile),
+            if (messageFile.getProtocol == "file") messageFile.getPath
+            else messageFile.toString
+          )
```
Instead of displaying a URI like `file:///home/user/my_project/conf/messages` in the dev error page, we should just display `/home/user/my_project/conf/messages` - also this way, when `play.editor` is set, the user can actually open the file (not possible with urls, Intellij doesn't do anything)

mkurz commented on dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/dev-mode-compile-and-config-error-source/Foo.java.bad:
```diff
@@ -0,0 +1,10 @@
+
+public class Foo {
+    public static int someNumber() {
+    }
```
Compilation will fail because of missing return.

marcospereira replied:
Let's add a comment then, just to make it explicit.

mkurz commented on dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/dev-mode-compile-and-config-error-source/HomeController.java.bad:
```diff
@@ -0,0 +1,21 @@
+    public HomeController(final SomeComponent someComponent) { }
+
+    public Result index() {
+    }
```
Compilation will fail because of missing return.

mkurz commented on dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/dev-mode-compile-and-config-error-source/SomeComponent.java.bad:
```diff
@@ -0,0 +1,9 @@
+
+  public SomeComponent() {
+    throw new RuntimeException("Houston, we've had a problem here.");
+  }
```
`SomeComponent` gets injected into the controller. The `RuntimeException` will be wrapped into a Guice `ProvisionException`.

marcospereira replied:
Add a comment about it too. This is not easy to spot if the person does not know about Guice.

mkurz commented on dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/dev-mode-compile-and-config-error-source/blacklist.txt:
```diff
@@ -0,0 +1,3 @@
+${BASE}
+${base}
+${
```
I don't want to see any of these words/chars in the error page.. I can see them now with Play 2.8.7 in combination with sbt 1.4
![image](https://user-images.githubusercontent.com/644927/108608110-fcfaa100-73c4-11eb-8477-b5d81da82d4c.png)

mkurz commented on dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/dev-mode-compile-and-config-error-source/index.scala.html.bad:
```diff
@@ -0,0 +1,4 @@
+@(first: String, second: String)
+first line
+@
```
A single `@` is invalid in twirl.

mkurz commented on dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/dev-mode-compile-and-config-error-source/routes.bad:
```diff
@@ -0,0 +1 @@
+GET                           controllers.HomeController.index()
```
Missing the path (slash)
```
GET              /             controllers.HomeController.index()
```

marcospereira replied:
Again, better to add a comment to make it easy to spot. :-)

mkurz commented on transport/server/play-server/src/main/scala/play/core/server/DevServerStart.scala:
```diff
@@ -256,6 +255,48 @@ final class DevServerStart(
+                            def sourceName = source._1.getAbsolutePath
+                          }
+                        )
+                        .getOrElse(UnexpectedException(message = Some(desc), unexpected = Some(other)))
```
FYI: An `ExceptionSource` will only show lines of the source file if there is a source. If the source does not exist it will show nothing, not even the exception... That's why I fall back via `map(source => ...).getOrElse` because an `UnexpectedException` will at least display the stacktrace of the exception (`UnexpectedException` basically is just a plain `PlayException`)

mkurz commented on core/play/src/main/scala/play/api/Configuration.scala:
```diff
@@ -145,20 +145,24 @@ object Configuration {
+          override def toString = "Configuration error: " + getMessage
+        }
+      })
+      .getOrElse(new PlayException("Configuration error", message, e.orNull))
```
FYI: When rendering the dev error page (see [here](https://github.com/playframework/playframework/blob/2.8.7/core/play/src/main/scala/views/defaultpages/devError.scala.html)) a `PlayException.ExceptionSource` will only show the lines of the source file if there is a source. If the source does not exist it will show just the heading (Like "Compilation error") and the message of exception, but not the stack trace of the exception... That's why I fall back via `map(....).getOrElse(new PlayException(...))` because a `PlayException` will at least display the stacktrace of the exception, which give the user more information.

mkurz commented on transport/server/play-server/src/main/scala/play/core/server/DevServerStart.scala:
```diff
@@ -256,6 +255,48 @@ final class DevServerStart(
                 lastState
 
+              // No binary dependency on play-guice
+              case e if e.getClass.getName == "com.google.inject.ProvisionException" =>
```
Inspired by https://github.com/playframework/playframework/blob/bf29c44758bee854257c0e9423af7498a3378f39/transport/server/play-server/src/main/scala/play/core/server/DevServerStart.scala#L251-L252 (added by Dale)

mkurz commented on core/play/src/main/scala/play/api/http/HttpErrorHandler.scala:
```diff
@@ -356,15 +356,36 @@ object HttpErrorHandlerExceptions {
-        def position   = null
-        def input      = source.map(_._1).map(f => PlayIO.readFileAsString(f.toPath)).orNull
-        def sourceName = source.map(_._1.getAbsolutePath).orNull
+    case e if e.getClass.getName == "com.google.inject.ProvisionException" => // No binary dependency on play-guice
```
Similiar like below: https://github.com/playframework/playframework/pull/10711/files#r579708394
We have to check for `ProvisionException` here in HttpErrorHandler and in DevServerStart:
In DevServerStart it is necessary to catch exceptions when initializing the built-in Guice modules (like messages api, to fix problems like https://github.com/playframework/playframework/issues/10710#issue-810030125)
Here in HttpErrorHandler we have to handle `ProvisionException` as well for problems initializing Guice caused by user code (like https://github.com/playframework/playframework/issues/10710#issuecomment-781217827). This get's testet in the scripted test with `SomeComponent`, just see below.

mkurz commented on core/play/src/main/scala/play/api/http/HttpErrorHandler.scala:
```diff
@@ -356,15 +356,36 @@ object HttpErrorHandlerExceptions {
+          def sourceName = source._1.getAbsolutePath
+        }
+      )
+      .getOrElse(new PlayException("Execution exception", desc, throwable))
```
Again the fallback to `PlayException` to at least display the user the exception stack trace instead of almost nothing (just heading an exceptoin message)

mkurz commented on dev-mode/sbt-plugin/src/main/scala/play/sbt/PlayExceptions.scala:
```diff
@@ -28,7 +28,17 @@ object PlayExceptions {
+        // A file outside of the base project folder or using sbt <1.4
+        sourceFile
+      }
+    }
```
This actually is the fix needed to support sbt 1.4+ in Play :tada:

mkurz commented:
@octonato Maybe you can review this pull request here? Because you already reviewed #10648 (the other sbt pr) and know how this virtual file stuff works. Thanks!

marcospereira reviewed (commented):
Overall, this looks very good to me. Thanks for adding such extensive tests. :-)

I made some very minor comments. Most of the comments you made here to explain how things are failing should go into the code. But besides that, this looks ready to be merged.

mkurz commented:
@marcospereira Comments added. I also enhanced the existing scripted test added in #10648 with a check I added in this PR.

Updated the backport #10712 as well.

Looking forward to your approval of both :wink:

octonato reviewed (approved):
Thanks @mkurz 

Lot's of work on this. Wow!
